File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/TopInterview2Round/maxPathSum_124.java
package algo.tzashinorpu.SpecialSubject.TopInterview2Round;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class maxPathSum_124 {
    int res = 0;

    public int maxPathSum(TreeNode root) {
        res = root.val;
        helper(root);
        return res;
    }

    private int helper(TreeNode root) {
        if (root == null) return 0;
        int l = Math.max(helper(root.left), 0);
        int r = Math.max(helper(root.right), 0);
        res = Math.max(res, l + r + root.val);
        return Math.max(l, r) + root.val;
    }
}

File: Clion/ZeroTrac/Tree/maxPathSum_124.cpp
//
// Created by TzashiNorpu on 10/18/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::maxPathSum(TreeNode *root) {
  int res = -10000;
  function<int(TreeNode *node)> dfs = [&](TreeNode *node) -> int {
      if (!node) return 0;
      int l = max(0,dfs(node->left));
      int r = max(0,dfs(node->right));
      res = max(res, l+r+node->val);
      return node->val+max(l,r);
  };
  dfs(root);
  return res;
}